### PR TITLE
Jormun: Context for all apis

### PR DESCRIPTION
**Pull request by azime**
for more informations, [See](https://jira.canaltp.fr/browse/NAVP-681)
Example for journeys API:
```json
{
....
  "context": {
    "car_direct_path": {
      "co2_emission": {
        "value": 50.23,
        "unit": ""
      }
    },
    "current_datetime": "20120614T080000",
    "timezone": "Europe/Paris"
  }
}
```

Example for ptref API:
```json
{
...
  "context": {
    "current_datetime": "20120614T080000",
    "timezone": "Europe/Paris"
  }
}
```

### CODE REVIEW

File: source/jormungandr/jormungandr/interfaces/v1/serializer/api.py [hunk 1]
```diff
@@ -193,6 +234,10 @@ class CoverageSerializer(NullableDictSerializer):
 
 class CoveragesSerializer(serpy.DictSerializer):
     regions = CoverageSerializer(many=True)
+    context = MethodField(schema_type=ContextSerializer(), display_none=False)
+
+    def get_context(self, obj):
+        return ContextSerializer(obj, display_none=False).data
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/api.py]: I don't think that's relevant here.

**Review by TeXitoi — APPROVED** [review]

File: source/jormungandr/jormungandr/interfaces/v1/fields.py [hunk 2]
```diff
@@ -936,3 +961,45 @@ def output(self, key, obj):
...
+        "networks": network,
+        "disruptions": None,
+        "contributors": contributor,
+        "datasets": dataset,
+    }
+
+    if collection_name not in map_collection:
+        raise CollectionException
+
+    if map_collection[collection_name]:
```
Comment (xlqian) [inline source/jormungandr/jormungandr/interfaces/v1/fields.py]: hmmm, you do the search twice: once on L999, once here, use `get` :)

File: source/jormungandr/jormungandr/interfaces/v1/fields.py [hunk 3]
```diff
@@ -936,3 +961,45 @@ def output(self, key, obj):
...
+        return [(collection_name,
+                 NonNullList(fields.Nested(map_collection[collection_name], display_null=False)))] + common_collection
+    return [] + common_collection
```
Comment (xlqian) [inline source/jormungandr/jormungandr/interfaces/v1/fields.py]: you can return `common_collection`, no?

File: source/jormungandr/jormungandr/interfaces/v1/serializer/api.py [hunk 4]
```diff
@@ -34,15 +34,44 @@
...
+from jormungandr.interfaces.v1.serializer.journey import TicketSerializer, JourneySerializer
 import serpy
 
 from jormungandr.interfaces.v1.serializer.jsonschema.fields import Field, MethodField
 from jormungandr.interfaces.v1.serializer.time import DateTimeDictField
+from jormungandr.utils import get_current_datetime_str, get_timezone_str
+
+
+class CO2Serializer(PbNestedSerializer):
+    co2_emission = AmountSerializer(attr='car_co2_emission', display_none=False)
```
Comment (xlqian) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/api.py:50]: can you do `car_co2_emission = AmountSerializer(display_none=False)`?

File: source/jormungandr/jormungandr/interfaces/v1/serializer/api.py [hunk 5]
```diff
@@ -135,13 +168,21 @@ class PlacesSerializer(serpy.Serializer):
...
+    def get_context(self, obj):
+        return ContextSerializer(obj, display_none=False).data
 
 
 class PtObjectsSerializer(serpy.Serializer):
     error = ErrorSerializer(display_none=False)
     feed_publishers = FeedPublisherSerializer(many=True, display_none=True)
     disruptions = pt.DisruptionSerializer(attr='impacts', many=True, display_none=True)
     pt_objects = pt.PtObjectSerializer(many=True, attr='places')
+    context = MethodField(schema_type=ContextSerializer(), display_none=False)
```
Comment (xlqian) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/api.py:181]: add a comment explain why not using ContextSerializer directly

File: source/jormungandr/tests/ptref_tests.py [hunk 6]
```diff
@@ -112,7 +112,7 @@ def test_vj_default_depth(self):
         assert (feed_publisher["name"] == "canal tp")
         assert (feed_publisher["license"] == "ODBL")
         assert (feed_publisher["url"] == "www.canaltp.fr")
-
+        self.check_context(response)
```
Comment (xlqian) [inline source/jormungandr/tests/ptref_tests.py:115]: :+1:

File: source/jormungandr/jormungandr/utils.py [hunk 7]
```diff
@@ -467,3 +479,17 @@ def make_namedtuple(typename, *fields, **fields_with_default):
...
     T.__new__.__defaults__ = tuple(fields_with_default.values())
     return T
+
+
+def get_timezone_str(default='Africa/Abidjan'):
+    try:
+        timezone = get_timezone()
+    except TechnicalError:
+        pass
+    return timezone.zone if timezone else default
```
Comment (xlqian) [inline source/jormungandr/jormungandr/utils.py]: i think you can do 
```python
try:
    timezone = get_timezone()
except TechnicalError:
    return default:
else:
    return timezone.zone
```
TO BE TESTED

File: source/jormungandr/jormungandr/interfaces/v1/fields.py (hunk 2 above)
Comment (azime) [reply source/jormungandr/jormungandr/interfaces/v1/fields.py]: Done

File: source/jormungandr/jormungandr/interfaces/v1/fields.py (hunk 3 above)
Comment (azime) [reply source/jormungandr/jormungandr/interfaces/v1/fields.py]: Done

File: source/jormungandr/jormungandr/utils.py (hunk 7 above)
Comment (azime) [reply source/jormungandr/jormungandr/utils.py]: Done

File: source/jormungandr/jormungandr/interfaces/v1/serializer/api.py (hunk 4 above)
Comment (azime) [reply source/jormungandr/jormungandr/interfaces/v1/serializer/api.py:50]: the integration will not work anymore

File: source/jormungandr/jormungandr/interfaces/v1/serializer/api.py (hunk 5 above)
Comment (azime) [reply source/jormungandr/jormungandr/interfaces/v1/serializer/api.py:181]: Done

**Review by xlqian — APPROVED** [review]

File: source/jormungandr/jormungandr/interfaces/v1/serializer/api.py [hunk 8]
```diff
@@ -34,15 +34,44 @@
...
+class ContextSerializer(PbNestedSerializer):
+    car_direct_path = MethodField(schema_type=CO2Serializer(), display_none=False)
+    current_datetime = MethodField(schema_type=str,
+                                   display_none=False,
+                                   description='The datetime considered as "now". Used for debug, default is '
+                                               'the moment of the request. It will mainly change the output '
+                                               'of the disruptions.')
+    timezone = MethodField(schema_type=str,
+                           display_none=False,
+                           description='timezone of region, default value Africa/Abidjan')
```
Comment (pbougue) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/api.py]: I'd add: Africa/Abidjan` (UTC)`

File: source/jormungandr/jormungandr/interfaces/v1/serializer/api.py [hunk 9]
```diff
@@ -34,15 +34,44 @@
...
+
+class CO2Serializer(PbNestedSerializer):
+    co2_emission = AmountSerializer(attr='car_co2_emission', display_none=False)
+
+
+class ContextSerializer(PbNestedSerializer):
+    car_direct_path = MethodField(schema_type=CO2Serializer(), display_none=False)
+    current_datetime = MethodField(schema_type=str,
+                                   display_none=False,
+                                   description='The datetime considered as "now". Used for debug, default is '
```
Comment (pbougue) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/api.py]: Please change it to: The datetime of the request (considered as "now").
The rest of it is unrelated.

File: source/jormungandr/jormungandr/interfaces/v1/serializer/api.py [hunk 10]
```diff
@@ -51,6 +80,11 @@ class PTReferentialSerializer(serpy.Serializer):
     feed_publishers = FeedPublisherSerializer(many=True, display_none=True)
     disruptions = pt.DisruptionSerializer(attr='impacts', many=True, display_none=True)
     notes = DescribedField(schema_type=NoteSerializer(many=True))
+    #ContextSerializer can not be used directly because context does not exist in protobuff
```
Comment (pbougue) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/api.py]: typo: protobuf (one f)

**Review by pbougue — COMMENTED** [review]
We should also add it to the doc as it's an important feature.

File: source/jormungandr/jormungandr/interfaces/v1/serializer/api.py (hunk 10 above)
Comment (azime) [reply source/jormungandr/jormungandr/interfaces/v1/serializer/api.py]: Done

File: source/jormungandr/jormungandr/interfaces/v1/serializer/api.py (hunk 8 above)
Comment (azime) [reply source/jormungandr/jormungandr/interfaces/v1/serializer/api.py]: Done

File: source/jormungandr/jormungandr/interfaces/v1/serializer/api.py (hunk 9 above)
Comment (azime) [reply source/jormungandr/jormungandr/interfaces/v1/serializer/api.py]: Done

**Pull request merged**